app/plugin.go

```diff
@@ -489,24 +495,64 @@ func (a *App) getRemotePlugins(filter *model.MarketplacePluginFilter) ([]*model.
 			continue
 		}
 
-		result = append(result, &model.MarketplacePlugin{BaseMarketplacePlugin: p})
+		result[p.Manifest.Id] = &model.MarketplacePlugin{BaseMarketplacePlugin: p}
 	}
 
 	return result, nil
 }
 
-// addLocalPlugins returns a merged list of locally installed and remote marketplace plugins.
-func (a *App) addLocalPlugins(remoteMarketplacePlugins []*model.MarketplacePlugin) ([]*model.MarketplacePlugin, *model.AppError) {
-	allPlugins := map[string]*model.MarketplacePlugin{}
+// mergePrepackagePlugins returns a merged list of pre-packaged plugins and remote marketplace plugins.
+func (a *App) mergePrepackagePlugins(remoteMarketplacePlugins map[string]*model.MarketplacePlugin) (map[string]*model.MarketplacePlugin, *model.AppError) {
+	allPlugins := remoteMarketplacePlugins
 
-	for _, p := range remoteMarketplacePlugins {
-		if p.Manifest == nil {
+	pluginsEnvironment := a.GetPluginsEnvironment()
+	if pluginsEnvironment == nil {
+		return nil, model.NewAppError("GetMarketplacePlugins", "app.plugin.config.app_error", nil, "", http.StatusInternalServerError)
+	}
+
+	for _, prepackaged := range pluginsEnvironment.PrepackagedPlugins() {
+		if prepackaged.Manifest == nil {
+			continue
+		}
+
+		prepackagedMarketplace := &model.MarketplacePlugin{
+			BaseMarketplacePlugin: &model.BaseMarketplacePlugin{
+				Manifest: prepackaged.Manifest,
+			},
+		}
+
+		// If not available in marketplace, add the prepackaged
+		if allPlugins[prepackaged.Manifest.Id] == nil {
+			allPlugins[prepackaged.Manifest.Id] = prepackagedMarketplace
 			continue
 		}
-		allPlugins[p.Manifest.Id] = p
+
+		// If available in the markteplace, only overwrite if newer.
+		var prepackagedVersion, marketplaceVersion semver.Version
+
+		prepackagedVersion, err := semver.Parse(prepackaged.Manifest.Version)
+		if err != nil {
+			return nil, model.NewAppError("GetMarketplacePlugins", "app.plugin.invalid_version.app_error", nil, "", http.StatusBadRequest)
+		}
+
+		marketplacePlugin := allPlugins[prepackaged.Manifest.Id]
+		marketplaceVersion, err = semver.Parse(marketplacePlugin.Manifest.Version)
+		if err != nil {
+			return nil, model.NewAppError("GetMarketplacePlugins", "app.plugin.invalid_version.app_error", nil, "", http.StatusBadRequest)
+		}
+
+		if prepackagedVersion.GT(marketplaceVersion) {
+			allPlugins[prepackaged.Manifest.Id] = prepackagedMarketplace
+		}
 	}
 
-	// Include all other installed plugins.
+	return allPlugins, nil
```

Comment:
Note that `allPlugins == remoteMarketplacePlugins`, so we've actually already modified the input parameter here, making the act of returning it redundant, and perhaps confusing if the expectation was that we would work on a copy instead.

I'd propose either working on a copy, or just not returning the value and having the current semantics be explicit.

Reply:
Updated.